includes/handlers/class-webmention-handler-meta.php

```diff
@@ -0,0 +1,208 @@
+<?php
+/**
+ * Class for webmention parsing using META tags.
+ */
+class Webmention_Handler_Meta extends Webmention_Handler_Base {
+
+	/**
+	 * Takes a request object and parses it.
+	 *
+	 * @param Webmention_Request $request Request Object.
+	 * @param Webmention_Item $item A Parsed Item. If null, a new one will be created.
+	 * @return WP_Error|true Return error or true if successful.
+	 */
+	public function parse( $request, $item = null ) {
+		$this->set_webmention_item( $item );
+
+		$dom   = clone $request->get_domdocument();
+		$xpath = new DOMXPath( $dom );
+
+		$meta = array();
+		// Look for OGP properties
+		foreach ( $xpath->query( '//meta[(@name or @property or @itemprop) and @content]' ) as $tag ) {
+			$meta_name = $tag->getAttribute( 'property' );
+			if ( ! $meta_name ) {
+				$meta_name = $tag->getAttribute( 'name' );
+			}
+			if ( ! $meta_name ) {
+				$meta_name = $tag->getAttribute( 'itemprop' );
+			}
+			$meta_value = $tag->getAttribute( 'content' );
+			// Sanity check. $key is usually things like 'title', 'description', 'keywords', etc.
+			if ( strlen( $meta_name ) > 200 ) {
+				continue;
+			}
+			$meta = self::add_property( $meta, $meta_name, $meta_value );
+		}
+
+		$meta['title'] = trim( $xpath->query( '//title' )->item( 0 )->textContent );
+		$meta          = self::parse_ogp( $meta );
+		if ( isset( $meta['og'] ) ) {
+			$meta['og'] = self::parse_ogp( $meta['og'] );
+		}
+
+		// Set raw data.
+		$this->webmention_item->set__raw( $meta );
+		$this->webmention_item->set__response_type = 'mention';
+
+		$item = $this->ogp( $meta );
+		$item = array_merge( $item, $this->dublin_core( $meta ) );
+
+		// If Site Name is not set use domain name less www
+		if ( ! isset( $item['_site_name'] ) && isset( $item['url'] ) ) {
+			$item['_site_name'] = preg_replace( '/^www\./', '', wp_parse_url( $item['url'], PHP_URL_HOST ) );
+		}
+
+		if ( ! isset( $item['name'] ) ) {
+			$item['name'] = $meta['title'];
+		}
+
+		if ( isset( $meta['citation_date'] ) && ! isset( $item['published'] ) ) {
+			$item['published'] = new DateTimeImmutable( $meta['citation_date'] );
+		} elseif ( isset( $meta['datePublished'] ) ) {
+			$item['published'] = new DateTimeImmutable( $meta['datePublished'] );
+		}
+
+		if ( ! isset( $item['author'] ) ) {
+			$item['author'] = array( 'name' => $meta['author'] );
+		}
+
+		$item = array_filter( $item );
+		foreach ( $item as $key => $value ) {
```

Comment:
why such a "complicated" usage of different array formats instead of directly add them to the item-object? It is very hard to verify it like this. It feels like it does things twice.

Reply:
I was trying to make it simpler to read, but didn't quite work out. I can redo it to touch the object...